### ISSUE TITLE
Adding in correlation metrics

### ISSUE THREAD

**Issue by mortonjt** [biocore/scikit-bio#1118]
I'm thinking about adding in [Lovell's Proportionality](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075) metric into `skbio.stats.composition`.

Basically, it's an awesome correlation metric suited for relative abundance data - and pretty much kills other correlation metrics such as Pearson and Spearman.

If you guys think this is a suitable addition for `skbio`, the following PR will include two publicly available functions
1. Cythonized calculation of the Aitchison variation matrix
2. Lovell's proportionality metric function

Also, note that the first function is required for Sparcc, so this would be a good stepping stone if we decide to implement that.

**Comment by ebolyen** [comment]
That's a cool paper! :+1: for implementing. Granted it would be ideal if this was in `statsmodels`, but given that this correlation metric was designed in the context of bioinformatics, I think it has a place here. 

I think it would be good to work out what we call this functions in this issue.

Would we call the propertionality metric: `lovell_proportionality` for example?

**Comment by jairideout** [comment]
This looks really interesting! I agree scikit-bio is a good home for this, and since it is such a new metric a Python implementation probably doesn't exist yet.

:+1: for `lovell_proportionality`.

**Comment by ElDeveloper** [comment]
> ... since it is such a new metric a Python implementation probably doesn't exist yet.

Sad sad world we live in, seems like the publication is accompanied by R code only `:L`

**Comment by mortonjt** [comment]
I like `lovell_proportionality`!

Note that there are actually two different proportionality metrics

1. Measure of positive proportionality (the author's preferred metric)
2. Measure of positive/negative proportionality

So I'm thinking about having this function formatted something like
`lovell_proportionality(x, two_sided=True)`
Where if `two_sided=True`, then it'll test for both positive and negative proportionality.

Also @ElDeveloper , don't get too down.  There is only 1 implementation in R, and it is neither fast or very useable.

**Comment by jairideout** [comment]
`two_sided` sounds like an alternative hypothesis to me (e.g., see `mantel`'s `alternative` parameter, which accepts `'two-sided'`, `'greater'`, or `'less'`). Is this a similar concept or is it a different test statistic that ranges from -1 to +1 instead of 0 to +1? I'm assuming this is the equation described in section "Alternative measures of proportionality".

It may be better to first implement the author's preferred metric (and make this the default). The alternative metric seems to be almost an afterthought (and probably warrants its own paper demonstrating its applicability).

**Comment by mortonjt** [comment]
:+1: That sounds like a plan to me.  We can implement the author's preferred metric and perhaps add in the alternative metric in later.  Probably will want to have that extra parameter though and perhaps raise a `NotImplementedError`.  One thing to note is that there is no published hypothesis test with this proportionality metric yet - that is still active research as far as I know.

Yeah, its a different statistic that ranges from -1 to +1 that is outlined in the "Alternative measures of proportionality".  The author's preferred statistic is actually unbounded above - it ranges from 0 to infinity.

**Comment by ebolyen** [comment]
@mortonjt We can add the extra parameter later if needed, that would be a non-breaking change.

**Comment by jairideout** [comment]
Thanks for the info regarding the test statistic range. Since this is an entirely different test statistic (with different interpretation) do you think it would be better to have a separate function instead of a parameter? (in the future of course) This would be similar to how there are two functions for Pearson and Spearman correlation.

**Comment by mortonjt** [comment]
Having a separate function for the other statistic is definitely a possibility.  I'm just not sure what to call it haha

**Comment by jairideout** [comment]
Yeah I'm not able to think of a good name either... The preferred statistic measures "goodness-of-fit to proportionality" while the alternative is a "proportionality correlation coefficient". @ebolyen any ideas?

**Comment by jolespin** [comment]
> At least three measures of proportionality exist. The first, ϕ, ranges from [0,inf] with 0 indicating perfect proportionality (Lovell et al., 2015). Its definition adjusts the VLR (in the numerator) by the variance of one of the log-ratio transformed features in that pair (in the denominator). The use of only one feature variance in the adjustment makes ϕ asymmetric. The second, ϕs, also ranges from [0,inf] with 0 indicating perfect proportionality, but has a natural symmetry (Quinn et al., 2017b). Its definition adjusts the VLR by the variance of the log-product of the two features. The third, ρp, like correlation, takes on values from [–1, 1], where a value of 1 indicates perfect proportionality (Erb and Notredame, 2016). Its definition adjusts the VLR by the sum of the variances of the log-ratio transformed features in that pair (as subtracted from the value 1). Thus, ρp is also symmetric.
https://academic.oup.com/bioinformatics/article/34/16/2870/4956011

> #'  One can think of phi as
> #' 	analogous to a distance matrix, except that it has no symmetry unless forced.
> #' 	One can think of rho as
> #' 	analogous to a correlation matrix.
> #' 	One can think of phs as
> #' 	either a naturally symmetric variant of phi or a monotonic variant of rho.
> #' 	Also, one can use \code{corr}
> #' 	to calculate correlation from log-ratio transformed data.
https://github.com/tpq/propr/blob/2bd7c44bf59eaac6b4d329d38afd40ac83e2089a/R/1-propr.R#L25

What about the following: 

```
from typing import Union
def lovell_proportionality(data:Union[np.array, pd.DataFrame], metric:str="phi"):
    """
    metric: {phi, phs, rho}
    """
   return proportionality_matrix
```

Has there been any ideas for implementing this?  I will have to do this soon and I'm trying to avoid making an `rpy2` wrapper because there have been [some compatibility issues](https://github.com/rpy2/rpy2/issues/690) with the new version and, as you know, `propr` is quite frustrating...  As mentioned, `propr` doesn't appear to be very efficient https://github.com/tpq/propr/blob/master/R/1-propr.R. 

Also, any thought on implementing variance log ratio (VLR)?

**Comment by mortonjt** [comment]
I have implemented the first two metrics (see [here](https://github.com/mortonjt/benchmarkThis/blob/master/src/correlation.py#L10)). You are welcome to use them.

I haven't used VLR, so I can't comment on those - but I think having the first metrics in would be preferable to have in first before adding in VLR.

**Comment by jolespin** [comment]
@mortonjt Thanks for linking out to that!  It makes it so much easier to understand looking at the code than at the equations in the paper.  

Are there still plans to release this in some form in skbio?

**Comment by mortonjt** [comment]
Hopefully.  Most of the maintainers are pretty decked with responsibilities due to COVID recently. 
I don't know what the timeline for reviewing currently is.

**Comment by jolespin** [comment]
I feel you, very interesting times right now.  I have some papers I'm working on but once they are submitted I will try to take a shot at implementing this.  Whatever pieces are missing that are useful from the `propr` package.  Hopefully more (computational) biologists start using Python more because `rpy2` is pretty buggy and pandas doesn't seem interested in maintaining [interoperability](https://github.com/pandas-dev/pandas/issues/33754).

**Comment by pirovc** [comment]
Thanks for sharing the code @mortonjt . I would also be interested on the implementation of VLR

**Comment by mortonjt** [comment]
Awesome! Feel free to submit a pull request and I’ll be happy to review it
:)

**Comment by jolespin** [comment]
I've developed this package and would be happy to integrate it into scikit-bio if you feel that it would fit in there. 

https://github.com/jolespin/compositional

**Comment by pirovc** [comment]
Thanks for sharing @jolespin, looks like you already did it all so integration should be straightforward and jugding by this issue, proportionality would be a good fit for scikit-bio. @mortonjt what would be the best approach to the PR? I guess implementing the functions: rho, phi, phs and vlr to the skbio.stats.composition module?

**Comment by ElDeveloper** [comment]
My suggestion would be to start with one of the functions, that way the pull request will be easier to review and we'll have more flexibility to make big changes without having to rewrite a lot of code.

**Comment by jolespin** [comment]
Should we start with:
https://github.com/jolespin/compositional/blob/3fb1043d1b1e2140add7738c0d8ac5d033e74c7a/compositional/compositional.py#L329

Since it needs no adjustments to use w/ scikit-bio?

https://github.com/jolespin/compositional/blob/3fb1043d1b1e2140add7738c0d8ac5d033e74c7a/compositional/compositional.py#L369

https://github.com/jolespin/compositional/blob/3fb1043d1b1e2140add7738c0d8ac5d033e74c7a/compositional/compositional.py#L427

Will be simple adjustments.  Just need to replace my `transform_xlr` w/ `clr` from scikit-bio.

**Comment by jolespin** [comment]
I'm assuming we don't want any checks in this right?  Removing this type of code: 
```python
#     if checks:
#         # Checks
#         n_dimensions = len(X.shape)
#         assert n_dimensions in {2}, "`X` must be 2D"
#         assert np.all(X >= 0), "`X` cannot contain negative values because of log-transformation step."
        
#         n_zeros = np.any(X == 0).flatten().sum()
#         if n_zeros:
#             raise Exception("N={} zeros detected in `X`.  Either preprocess or add pseudocounts.".format(n_zeros))    

```


What are your thoughts on having the `ids` parameter function in the following way.  If `mat` is a `pd.DataFrame` it automatically grabs the component names BUT if `ids` is provided then that is used instead when creating the `DistanceMatrix`.  This will alleviate the user from an extra step of providing `ids`.

I can remove this feature if this doesn't match the API well enough but I thought I would propose it anyways since I would likely write wrappers to do this externally. 


```python
import numpy as np
import pandas as pd
from skbio.util._decorator import experimental
from skbio.stats.distance import DissimilarityMatrix, DistanceMatrix
from typing import List, Union


@experimental(as_of="0.5.x")
def vlr(mat:Union[np.array, pd.DataFrame], ids:list=None, ddof:int=1) -> DistanceMatrix:#, checks=False): 
    """
    # Description
    Pairwise variance log-ratio
    
    # Parameters
        * X: pd.DataFrame or 2D np.array
        
    Adapted from the following source:
    * https://github.com/tpq/propr
    ddof=1 for compatibility with propr package in R
    """
    
    # Identifiers
    components = None
    if isinstance(mat, pd.DataFrame):
        components = mat.columns
        mat = mat.values
    if ids is None:
        ids = components
    
    # Variance Log Ratio
    X = mat.astype("float64")
    X_log = np.log(X)
    covariance = np.cov(X_log.T, ddof=ddof) 
    diagonal = np.diagonal(covariance)
    vlr_data = -2*covariance + diagonal[:,np.newaxis] + diagonal 

    return DistanceMatrix(vlr_data, ids=ids)
```

Note, I still need to update the doc string so it matches the rest of the skbio functions.

**Comment by mortonjt** [comment]
If you run `closure` that will automatically handle the checks.

@ebolyen , @ElDeveloper do you have strong opinions on the dataframe typing?  I feel that if we do it here, we'd also need to change all of the functions (which I am not opposed to).

Note that R is GPL'ed, so we may have some problems here if you directly adapted it from propr.  
Did you adapt it from the R code, or did you adapt it from the math and/or documentation?

**Comment by jolespin** [comment]
@mortonjt a bit of both.  I made sure that the output was the same between what I was doing in Python and the output of propr in R.  There are some non-optimized versions that made it a bit easier to translate the math to functions: https://github.com/tpq/propr/blob/master/R/9-deprecated.R It was fairly straightforward, just had to match up the `ddof` in most cases.

**Comment by mortonjt** [comment]
I think there could be GPL issues here that may not make this pull request BSD compatible.

Could you include the original citation that described the vlr?  This isn't the Aitchison total variational matrix is it?

I bit confused about the error - I can't comment on the usage of closure without the full code and example data.
I'd refer to the [contributing docs](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md) for more information on how to make a pull request.

**Comment by jolespin** [comment]
Here is the original function: 
```
proprVLR <- function(X, check = FALSE){

  if(check){

    if(any(X <= 0)) stop("non-positive values found")
    if(any(is.na(X))) stop("NA values found")
  }

  X <- log(X)
  X <- stats::var(X)
  X.diag <- diag(X)

  for(col in 1:ncol(X)){

    X[, col] <- -2 * X[, col] + X.diag + X.diag[col]
  }

  return(X)
}
```

I can remove that statement since it's not porting the code line by line. 

The original implementation is here: 
https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

The reason why I referenced `propr` is because I made sure the output matched `propr` which has been a peer-reviewed implementation of Lovell et al. 2015. 

Can I remove the reference to `propr` and add in the original Lovell et al. 2015?

**Comment by mortonjt** [comment]
If you can transcribe the function directly from the paper yes and it is ok to compare against the propr output.  
But it'll be problematic if the python code was transcribed from propr.

**Comment by mortonjt** [comment]
Alright, I'm looking at the VLR functionality more carefully and I think I know where the confusion is.

Variance log ratio is defined below
![image](https://user-images.githubusercontent.com/4184797/131428779-87823835-5f58-4509-943f-2242f74dbe18.png)

This equation is actually not completely correct -- the logarithms are there for illustration purposes, but you still need to ensure that some form of scale invariance is satisfied.  The easiest way to do this is to perform a CLR transform before even computing the variances, as suggested by the authors.

![image](https://user-images.githubusercontent.com/4184797/131428887-e7e5bc30-71ca-4efe-8be6-a06db3734660.png)

So the proprVLR function is actually incorrect, it should be using a CLR instead of raw proportions.
This should be ok, I think we can just move forward with the clr implementation so long as we convince ourselves that this correct.

Also note that this is different than the Aitchison variation matrix (see top of page 40 [here](http://www.leg.ufpr.br/lib/exe/fetch.php/pessoais:abtmartins:a_concise_guide_to_compositional_data_analysis.pdf)), which is an even simpler version that what is proposed in this issue.

There is also the [B-associations paper](https://www.researchgate.net/publication/322809225_Linear_Association_in_Compositional_Data_Analysis) that proposes a couple of additional means to do hypothesis testing to proportionality-like methods.

Not saying that we need to incorporate all of these metrics in, but pointing out that there are many favors of proportionality at this point, and we should be careful with the notation, even for published and widely used packages.

**Comment by pirovc** [comment]
as @ElDeveloper suggested and I sketched the code in a post above, wouldn't it be better to have the functions isolated and not as pairwise? we can check the performance later and add both if there's a big difference. +1 to @mortonjt suggestion of having the PR open so we can see the full code and suggest changes directly

**Comment by mortonjt** [comment]
I don't think it is unreasonable to include both `vlr` and `pairwise_vlr`, since they are both simple functions.  One benefit of having the `pairwise_vlr` is that we can leverage `DistanceMatrix` to determine if it is a valid matrix.  On the other hand, just having `vlr` as a barebones pull request is also not unreasonable, since we could use `DistanceMatrix.from_iterable` to recreate the full matrix (albeit its probably going to be a bit slower).  I have a slight preference for the former, but am ok with either type of PR.

**Comment by jolespin** [comment]
@mortonjt I had a question regarding splitting pairwise and singleton calculations.  If we decide that `CLR` instead of a `log` transform is the correct implementation, can we actually do a singleton calculation?  

Can `rho` and `phi` be calculated in a non-pairwise manner? 

I'm looking at my implementation of `rho` and it looks like we encounter a similar issue if we tried to implement in a singleton form (apologies if I'm butchering the pharsing): 

```python
# Pairwise rho proportionality
def pairwise_rho(X=None, reference_components=None, centroid="mean", interval_type="open", xlr=None, vlr=None):
    """
    # Description
    Pairwise proportionality `rho` (Erb et al. 2016)
    
    # Parameters
        * X: pd.DataFrame or 2D np.array of compositional data (rows=samples, columns=components)
        * reference_components: See `transform_xlr`.  Can also be `percentiles` for `transform_iqlr` or 'iqlr' string.
        * interval: 'open' = (a,b) and 'closed' = [a,b].  'open' is used by `propr` R package:
        * centroid: See `transform_xlr`
        * xlr: pd.DataFrame or 2D np.array of transformed compositional data (e.g. clr, iqlr) (must be used with `vlr` and not `X`)
        * vlr: pd.DataFrame or 2D np.array of variance log-ratios (must be used with `xlr` and not `X`)
        
    Citation:
    * https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

    """
    components = None
    # Compute xlr and vlr from X
    if X is not None:
        assert all(map(lambda x: x is None, [xlr, vlr])), "If `X` is not None then `xlr` and `vlr` cannot be provided."
        if isinstance(X, pd.DataFrame):
            components = X.columns
            X = X.values

        vlr = pairwise_vlr(X)
        if isinstance(reference_components, str):
            if reference_components.lower() == "iqlr":
                reference_components = (25,75)
        # Use percentiles
        if isinstance(reference_components, tuple):
            xlr = transform_iqlr(X,percentile_range=reference_components, centroid=centroid, interval_type=interval_type, zeros_ok=False)
        # Use CLR
        else:
            xlr = transform_xlr(X, reference_components=reference_components, centroid=centroid, zeros_ok=False)
    # Provide xlr and vlr
    else:
        assert all(map(lambda x: x is not None, [xlr,vlr])), "If `X` is None then `xlr` and `vlr` must be provided."
        assert type(xlr) is type(vlr), "`xlr` and `vlr` should be same type (i.e. pd.DataFrame, np.ndarray)"
        if isinstance(xlr, pd.DataFrame):
            assert np.all(xlr.columns == vlr.columns) & np.all(xlr.columns == vlr.index), "`xlr.columns` need to be the same as `vlr.index` and `vlr.columns`"
            components = xlr.columns
            xlr = xlr.values
            vlr = vlr.values
            
    # rho (Erb et al. 2016)
    n, m = xlr.shape
    variances = np.var(xlr, axis=0, ddof=1) # variances = np.var(X_xlr, axis=0, ddof=ddof)
    rhos = 1 - (vlr/np.add.outer(variances,variances))    
    if components is not None:
        rhos = pd.DataFrame(rhos, index=components, columns=components)
    return rhos
```

There's a lot of "bells and whistles" for flexibility on the above function it will be easy to standardize/simplify.

**Comment by mortonjt** [comment]
With `vlr` that can definitely be done on single features, so let's proceed with that first.

From previous benchmarks, I have seen that `phi` and `rho` are robust to feature selection, so it should be possible to specify this with just a pair of features.  But we'd need to confirm this (which we can do in the second pull request).

**Comment by jolespin** [comment]
So did we decide `clr` instead of just log transform is the way to go for `vlr`?

**Comment by mortonjt** [comment]
For VLR it doesn’t matter. So I’d just stick with log transform.

Another thing to think about is that VLR provides a unique opportunity to
drop the zeros. You can avoid computing the log ratio for a given pair in a
sample if one of them is zero and only compute the VLR on the remaining
samples.

**Comment by jolespin** [comment]
@pirovc is this the correct implementation for singleton calculations of `vlr`? 

```python
def vlr(x, y):
    return np.var(x - y, ddof=1)
```

Do we need to log transform?  The `var(x)` != `var(log(x))`. 
```python
>>> a = [1,2,3]
>>> np.var(a, ddof=1)
1.0
>>> np.var(np.log(a), ddof=1)
0.3086339881039915
```

@mortonjt do you mean something like this?

```
a = [0,1,2,3]
b = [1,2,3,4]

a_ = [1,2,3]
b_ = [2,3,4]

vlr(a_,b_)
```

If so, is that incorrect to do for `clr`?  That is, masking zeros with `nan` using `np.nanmean` instead of `np.mean` and then replacing `nan` with `0` after the `clr` calculation?

**Comment by mortonjt** [comment]
I’m assuming that the first implementation only takes logs as input.

RE masking correct - we have that be optional.
Yes, this is incorrect for CLR , but a similar approach can be done with
the robust CLR.
https://journals.asm.org/doi/full/10.1128/mSystems.00016-19
In which case the masking approach and robust clr will generate identical
results since the normalization constants will cancel

**Comment by jolespin** [comment]
That would make sense.  The reason I was asking was because `get_prop_matrix` takes a log function so I wasn't sure.

Thoughts on having something like `robust=True/False` for the `clr` function that uses the implementation in `deicode` w/o installing `deicode` separately (since `skbio` is a dependency) to make `skbio` a (mostly) "one stop shop" for CoDA in Python?

**Comment by mortonjt** [comment]
I’d be in favor of having a robust flag, but I don’t think we need robust
clr for this. There are masked log and masked cov numpy functions we can
use.

I’m ok with having more deicode functions here, but that’s a separate
discussion.

**Comment by jolespin** [comment]
Good call, one step at a time. I'll open a separate PR after we figure this one out. 

I've implemented the `vlr` for individual calculations and compared the pairwise versions.  The `np.log` calculations slow it down when it repeats. 

Below, `pairwise_vlr` is the original one that uses `np.cov` and `pairwise_vlr2` is the one that uses `scipy.spatial.distance` functions with `vlr` as the metric: 

```python
from scipy.spatial.distance import pdist, squareform


def vlr(x,y, ddof=1):
    x = np.log(x)
    y = np.log(y)
    return np.var(x - y, ddof=ddof)

@experimental(as_of="0.5.x")
def pairwise_vlr(mat:Union[np.array, pd.DataFrame], ids:list=None, ddof:int=1) -> DistanceMatrix:#, checks=False): 
    # Identifiers
    components = None
    if isinstance(mat, pd.DataFrame):
        components = mat.columns
        mat = mat.values
    if ids is None:
        ids = components
    
    # Variance Log Ratio
    X = mat.astype("float64")
    X_log = np.log(X)
    covariance = np.cov(X_log.T, ddof=ddof) 
    diagonal = np.diagonal(covariance)
    vlr_data = -2*covariance + diagonal[:,np.newaxis] + diagonal 
    
    return DistanceMatrix(vlr_data, ids=ids)

def pairwise_vlr2(mat:Union[np.array, pd.DataFrame], ids:list=None, ddof:int=1) -> DistanceMatrix:#, checks=False): 
    
    # Identifiers
    components = None
    if isinstance(mat, pd.DataFrame):
        components = mat.columns
        mat = mat.values
    if ids is None:
        ids = components
    
    # Variance Log Ratio
    X = mat.astype("float64")
    vlr_data = squareform(pdist(mat.T, metric=vlr))

    return DistanceMatrix(vlr_data, ids=ids)



a = pairwise_vlr(X_iris).to_series()
b = pairwise_vlr2(X_iris).to_series()
np.allclose(a,b)

# True
%timeit pairwise_vlr(X_iris)
# 143 µs ± 3.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit pairwise_vlr2(X_iris)
# 206 µs ± 7.74 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Can you elaborate on the masking function usage from `scipy`? 

I hard coded this version (I'm assuming this is much slower than what you are proposing) but it drops the pairwise speed down to 252 µs ± 16.4 µs per loop:
```
def vlr(x,y, ddof=1):
    mask = ((x == 0) + (y == 0)).astype(bool)
    x = np.log(x[~mask])
    y = np.log(y[~mask])
    return np.var(x - y, ddof=ddof)
```

Looks like `where` was introduced to masked functions in v1.20: 
https://numpy.org/doc/stable/reference/generated/numpy.ma.var.html

Not sure if this will introduce dependency issues.

**Comment by mortonjt** [comment]
Ideally, we'd only have `vlr` and `pairwise_vlr`, because like you said, `pairwise_vlr2` isn't going to be optimized.

Regarding the masking,  the masked version you proposed is implements the correct logic.  
But we could probably streamline it with some of the masked ops in numpy.

https://numpy.org/doc/stable/reference/maskedarray.generic.html#operations-on-masked-arrays
https://numpy.org/doc/stable/reference/generated/numpy.ma.cov.html

**Comment by jolespin** [comment]
Thanks for the suggestion on masked arrays.  I didn't know they could be use that way. 

**Can I create a PR with this or is it missing something?** 

```python
import numpy as np
import pandas as pd
from skbio.util._decorator import experimental
from skbio.stats.distance import DistanceMatrix
from typing import Union

@experimental(as_of="0.5.7")
def vlr(x,y, ddof=1, robust=False):
    r"""
    Calculates variance log ratio
    
    Parameters
    ----------
    x : array_like, float
       a 1-dimensional vector of proportions
    y : array_like, float
       a 1-dimensional vector of proportions
        
    ddof: int
        degrees of freedom
        
    robust: bool
        mask zeros at the cost of performance
       
    Returns
    -------
    float
         variance log ratio value
         
    Examples
    --------
    # No zeros
    >>> x = [1,2,3]
    >>> y = [5,8,13]
    >>> %timeit vlr(x,y)
    # 21.7 µs ± 979 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
    # 0.01277962183258352

    # Zeros without robust
    >>> x = [1,2,3,0]
    >>> y = [5,8,13,21]
    >>> %timeit vlr(x,y)
    # 27.5 µs ± 2.48 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
    # nan

    # Zeros with robust
    >>> x = [1,2,3,0]
    >>> y = [5,8,13,21]
    >>> %timeit vlr(x,y, robust=True)
    # 570 µs ± 21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
    # 0.01277962183258352
    
    References
    ----------
    .. [1] V. Lovell D, Pawlowsky-Glahn V, Egozcue JJ, Marguerat S, Bähler J (2015) 
           Proportionality: A Valid Alternative to Correlation for Relative Data. 
           PLoS Comput Biol 11(3): e1004075. https://doi.org/10.1371/journal.pcbi.1004075
    .. [2] Erb, I., Notredame, C. 
           How should we measure proportionality on relative gene expression data?. 
           Theory Biosci. 135, 21–36 (2016). https://doi.org/10.1007/s12064-015-0220-8
    """
    # Convert array_like to numpy array
    x = np.asarray(x)
    y = np.asarray(y)

    # Mask zeros
    if robust:
        x = np.ma.masked_array(x, mask=x == 0)
        y = np.ma.masked_array(y, mask=y == 0)
    
    # Log transformation
    x = np.log(x)
    y = np.log(y)
    
    #Variance log ratio
    return np.var(x - y, ddof=ddof)

@experimental(as_of="0.5.7")
def pairwise_vlr(mat:Union[np.array, pd.DataFrame], ids:list=None, ddof:int=1, robust=False) -> DistanceMatrix:
    r"""
    Performs pairwise variance log ratio transformation.
    
    Parameters
    ----------
    mat : array_like, float
       a matrix of proportions where
       rows = compositions and
       columns = components
       
    ids: array_like, str
        component names
        if pd.DataFrame is provided and `ids is None` then `mat.columns` are used for `ids`
        
    ddof: int
        degrees of freedom
        
    robust: bool
        mask zeros at the cost of performance
       
    Returns
    -------
    skbio.DistanceMatrix
         distance matrix of variance log ratio values
    Examples
    --------
    >>> # NumPy Array
    >>> mat = np.asarray([
        [1,2,3],
        [5,8,13],
        [21,34,55],
    ])
    >>> dism = pairwise_vlr(mat)
    >>> dism.redundant_form()
    # array([[0.        , 0.01576411, 0.00649553],
    #        [0.01576411, 0.        , 0.00202147],
    #        [0.00649553, 0.00202147, 0.        ]])

    >>> # Pandas DataFrame
    >>> df = pd.DataFrame(
        mat, 
        index=["sample_1", "sample_2", "sample_3"], 
        columns=["component_1", "component_2", "component_3"],
    )
    >>> dism = pairwise_vlr(df)
    >>> dism.to_data_frame()
    # component_1	component_2	component_3
    # component_1	0.000000	0.015764	0.006496
    # component_2	0.015764	0.000000	0.002021
    # component_3	0.006496	0.002021	0.000000
    
    References
    ----------
    .. [1] V. Lovell D, Pawlowsky-Glahn V, Egozcue JJ, Marguerat S, Bähler J (2015) 
           Proportionality: A Valid Alternative to Correlation for Relative Data. 
           PLoS Comput Biol 11(3): e1004075. https://doi.org/10.1371/journal.pcbi.1004075
    .. [2] Erb, I., Notredame, C. 
           How should we measure proportionality on relative gene expression data?. 
           Theory Biosci. 135, 21–36 (2016). https://doi.org/10.1007/s12064-015-0220-8
    """
    # Identifiers
    components = None
    if isinstance(mat, pd.DataFrame):
        components = mat.columns
        mat = mat.values
    if ids is None:
        ids = components
    
    # Mask zeros
    X = mat.astype(np.float64)
    if robust:
        X = np.ma.masked_array(X, mask=X == 0)
        
    # Log Transform
    X_log = np.log(X)
    
    # Variance Log Ratio
    covariance = np.cov(X_log.T, ddof=ddof) 
    diagonal = np.diagonal(covariance)
    vlr_data = -2*covariance + diagonal[:,np.newaxis] + diagonal 
    
    # Create distance matrix
    return DistanceMatrix(vlr_data, ids=ids)

```

**Also, how do we want `vlr` to handle zeros?** I thought of the following scenarios: 
1) An argument called `robust` that is slower but masks the zeros
2) An argument called `zero_method` like [scipy.stats.wilcoxon](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wilcoxon.html) where: 
2a. `nan` or `propogate` would return a `nan` (or something similar)
2b. `error` would return a `ValueError` (or a non-existent `skbio`-specific composotional error dealing with zeros)
2c. `robust` where it masks the zeros but does so at the expense of compute time

**Comment by mortonjt** [comment]
Yes please create a PR.   Don't handle zeros for now, we can always beef up
the PR as we go.

**Comment by ElDeveloper** [comment]
Looking good, so far. I would suggest including the MatJax for the equations in the docstrings. Regardless we can talk about that and other things in the PR. Thanks all.
